docs/proposals/high-availability.md

```diff
@@ -0,0 +1,40 @@
+# High Availability of Scheduling and Controller Components in Kubernetes
+This document serves as a proposal for high availability of the scheduler and controller components in kubernetes.  This proposal is intended to provide a simple High Availability api for kubernetes components with the potential to extend to services running on kubernetes.  Those services would be subject to their own constraints.
+
+## Design Options
+For complete reference see [this](https://www.ibm.com/developerworks/community/blogs/RohitShetty/entry/high_availability_cold_warm_hot?lang=en)
+
+1. Hot Standby: In this scenario, data and state are shared between the two components such that an immediate failure in one component causes the the standby deamon to take over exactly where the failed component had left off.  This would be an ideal solution for kubernetes, however it poses a series of challenges in the case of controllers where component-state is cached locally and not persisted in a transactional way to a storage facility.  This would also introduce additional load on the apiserver, which is not desireable.  As a result, we are **NOT** planning on this approach at this time. 
+
+2. **Warm Standby**: In this scenario there is only one active component acting as the master and additional components running but not providing service or responding to requests.  Data and state are not shared between the active and standby components.  When a failure occurs, the standby component that becomes the master must determine the current state of the system before resuming functionality.  This is the apprach that this proposal will leverage.
+
+3. Active-Active (Load Balanced): Clients can simply load-balance across any number of servers that are currently running.  Their general availability can be continuously updated, or published, such that load balancing only occurs across active participants.  This aspect of HA is outside of the scope of *this* proposal because there is already a partial implementation in the apiserver.
```

Comment:
> On Apr 26, 2015, at 2:00 AM, David Oppenheimer <notifications@github.com> wrote:
> 
> In docs/proposals/high-availability.md:
> 
> > @@ -0,0 +1,40 @@
> > +# High Availability of Scheduling and Controller Components in Kubernetes
> > +This document serves as a proposal for high availability of the scheduler and controller components in kubernetes.  This proposal is intended to provide a simple High Availability api for kubernetes components with the potential to extend to services running on kubernetes.  Those services would be subject to their own constraints.
> > +
> > +## Design Options
> > +For complete reference see [this](https://www.ibm.com/developerworks/community/blogs/RohitShetty/entry/high_availability_cold_warm_hot?lang=en)
> > +
> > +1. Hot Standby: In this scenario, data and state are shared between the two components such that an immediate failure in one component causes the the standby deamon to take over exactly where the failed component had left off.  This would be an ideal solution for kubernetes, however it poses a series of challenges in the case of controllers where component-state is cached locally and not persisted in a transactional way to a storage facility.  This would also introduce additional load on the apiserver, which is not desireable.  As a result, we are **NOT** planning on this approach at this time. 
> > +
> > +2. **Warm Standby**: In this scenario there is only one active component acting as the master and additional components running but not providing service or responding to requests.  Data and state are not shared between the active and standby components.  When a failure occurs, the standby component that becomes the master must determine the current state of the system before resuming functionality.  This is the apprach that this proposal will leverage.
> > +
> > +3. Active-Active (Load Balanced): Clients can simply load-balance across any number of servers that are currently running.  Their general availability can be continuously updated, or published, such that load balancing only occurs across active participants.  This aspect of HA is outside of the scope of *this* proposal because there is already a partial implementation in the apiserver.
> You didn't really say why you're not choosing this approach. For something like a scheduler it would probably be easy to use this approach (each of N simultaneously-running scheduler replicas is responsible for scheduling every Nth pod, or they compete to be the first to schedule each pod, or whatever).
> 
For truly level driven data flows the competitive model (with optimization to reduce contention) seems like the long term ideal.  I don't think any of our scheduler decisions should be perfect (just not bad), and done right at volume a few bad scheduling decisions based on inadequate info is no different than a flaky machine or bad network condition.
> The scheduler case might be more complicated; I haven't really thought about it. But I don't think "there is already a partial implementation in the apiserver" is the reason why you're not advocating this approach.
> 
> —
> Reply to this email directly or view it on GitHub.
>